autoprecompiles/src/constraint_optimizer.rs

```diff
@@ -93,9 +93,53 @@ fn solver_based_optimization<T: FieldElement, V: Clone + Ord + Hash + Display>(
     // does not increase.
     assert!(assignments.iter().all(|(_, expr)| expr.is_affine()));
     constraint_system.apply_substitutions(assignments);
+
+    // Now try to replace bus interaction fields that the solver knows to be constant
```

Comment:
Hmm, interesting.

So this is needed in cases where the bus interaction field is a complex expression referencing several variables, and we can't find the values of all the expressions (otherwise `constraint_system.apply_substitutions(assignments)` above would have already replaced the field by a constant), but we can solve for the value of the expression?

And it works because we'd have a "linearized" variable internally, but the assignment to it is filtered out by the `VarTransformation` "solver"?

I wonder if we can generalize this somehow. For example, if that same expression would appear in an algebraic constraint as a sub-expression, we'd also want to substitute it there.